core/shared/src/main/scala/zio/stm/ZSTM.scala

```diff
@@ -857,25 +905,64 @@ final class ZSTM[-R, +E, +A] private[stm] (
   def zip[R1 <: R, E1 >: E, B](that: => ZSTM[R1, E1, B]): ZSTM[R1, E1, (A, B)] =
     (self zipWith that)((a, b) => a -> b)
 
+  /**
+   * Named alias for `<&>`.
+   */
+  def zipPar[R1 <: R, E1 >: E, B](that: => ZSTM[R1, E1, B]): ZSTM[R1, E1, (A, B)] =
+    (self zipWithPar that)((a, b) => a -> b)
+
   /**
    * Named alias for `<*`.
    */
   def zipLeft[R1 <: R, E1 >: E, B](that: => ZSTM[R1, E1, B]): ZSTM[R1, E1, A] =
     (self zip that) map (_._1)
 
+  /**
+   * Named alias for `<&`.
+   */
+  def zipLeftPar[R1 <: R, E1 >: E, B](that: => ZSTM[R1, E1, B]): ZSTM[R1, E1, A] =
+    (self zipPar that) map (_._1)
+
   /**
    * Named alias for `*>`.
    */
   def zipRight[R1 <: R, E1 >: E, B](that: => ZSTM[R1, E1, B]): ZSTM[R1, E1, B] =
     (self zip that) map (_._2)
 
+  /**
+   * Named alias for `&>`.
+   */
+  def zipRightPar[R1 <: R, E1 >: E, B](that: => ZSTM[R1, E1, B]): ZSTM[R1, E1, B] =
+    (self zipPar that) map (_._2)
+
   /**
    * Sequentially zips this value with the specified one, combining the values
    * using the specified combiner function.
    */
   def zipWith[R1 <: R, E1 >: E, B, C](that: => ZSTM[R1, E1, B])(f: (A, B) => C): ZSTM[R1, E1, C] =
     self flatMap (a => that map (b => f(a, b)))
 
+  /**
+   * Sequentially zips this value with the specified one, combining the values
+   * using the specified combiner function and collecting errors.
+   */
+  def zipWithPar[R1 <: R, E1 >: E, B, C](that: => ZSTM[R1, E1, B])(f: (A, B) => C): ZSTM[R1, E1, C] =
+    self.continueWithM {
+      case TExit.Succeed(a) =>
+        that.continueWithM {
+          case TExit.Succeed(b) => ZSTM.succeed(f(a, b))
+          case TExit.Fail(c2)   => ZSTM.halt(c2)
+          case TExit.Retry      => ZSTM.retry
+        }
+      case TExit.Fail(c1) =>
+        that.continueWithM {
+          case TExit.Fail(c2) => ZSTM.halt(c1 ++ c2)
```

Comment:
Shouldn't it be `&&` here to indicate that these errors occurred "in parallel"?

Reply:
@adamgfraser The transactions occur sequentially so I thought `++` would be better suited in the usual case, leaving the user to use `&&` as part of an explicit return if they wanted to say something failed for two reasons at the same time. My motivation for these changes was that I wanted to fail adding an edge to a graph if the in vertex or the out vertex was missing, and wanted to be able to use `&&` to say that both the in vertex and out vertex were missing at the same time (i.e. it failed for both `Cause`s).

Reply:
Hey sorry for the delay on this. I think `&&` makes more sense here as we are doing them "in parallel" in the sense that we are doing both of them regardless of whether one fails and in particular I think that if both transactions fails we want  the cause from `l.zipPar(r)` to be the same as `r.zipPar(l)`, which is only true if the errors are "parallel". @mijicd do you have any thoughts on this?

Reply:
Order is important to my use case, which is the reason for raising this PR. 
```scala
//succeeds
graph.addVertex(sydney) &> graph.addVertex(melbourne) &> graph.addEdge(sydney, route1, melbourne) 

//fails at step 2 because melbourne isn't in the graph
graph.addVertex(sydney) &> graph.addEdge(sydney, route1, melbourne) &> graph.addVertex(melbourne) 

//fails at step 1 because melbourne isn't in the graph and (at the same time) because sydney isn't in the graph, and then later fails at step 3 because melbourne isn't in the graph
graph.addEdge(sydney, route1, melbourne) &> graph.addVertex(sydney) &> graph.addEdge(sydney, route2, melbourne) &>graph.addVertex(melbourne) 
```
I think this comes back to my previous concerns using the term `Par` for what are fundamentally sequential operations and aren't necessarily independent (`l` is always interpreted before `r`).

How about I swap to `&&` for `zipPar` and friends because that make sense if we're using the concept of parallel and will likely be useful for other use cases, then I can add some extra combinators that meet my needs for sequential collection of errors using `++`, say `zipSequential` & friends using variants of the symbolic operator `<#>` which isn't used anywhere else in the code base? `zipSeq` would probably be preferable but may cause confusion with the standard scala `Seq` trait. `<+>` would have nice parallels in the way the errors compose (`++` for `<+>` and `&&` for `<&>`) but that's already used for `orElseEither`.

Reply:
Got it. So then I think maybe what we need is the STM version of the `validateWith` operator on ZIO, which performs both effects in sequence and combines the causes with `++`. Then maybe we don't actually need the parallel versions. Sorry for the delay on this, just trying to think about how we fit it in with the tools we already have instead of doing something completely new.

Reply:
I think Par is misleading here. `validateWith` sounds reasonable.

Reply:
Yep, `validateWith` better reflects what I'm trying to do. I actually think we should keep the parallel versions because I do think they're useful for when people do want to run two unrelated transactions e.g. maintaining some sort of index of some other STM structure.

I'm actually using the `new ZSTM((journal, _, _, _) => {...}` trick in my graph library, but conceptually addEdge could look something like:
```scala
graph.getVertex(sydney) <&> graph.getVertex(melbourne) validateWith graph.underlying.addE(_.1, route1, _.2)
```
which I think would be pretty sweet!